File: JavaScript/153. Find Minimum in Rotated Sorted Array.js
// https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/
var findMin = function(nums) {
  let low = 0
  let high = nums.length - 1

  // Is there only one element or is array in sorted order?
  if (nums[0] <= nums[high]) {
    return nums[0]
  }

  while (low <= high) {
    const mid = Math.floor((low + high) / 2)

    if (nums[mid] < nums[mid - 1]) {
      return nums[mid]
    } else if (nums[mid] < nums[nums.length - 1]) {
      high = mid - 1
    } else {
      low = mid + 1
    }
  }

  return -1
};

File: Java/153. Find Minimum in Rotated Sorted Array.java
// https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/
class Solution {
    public int findMin(int[] nums) {
        int low = 0;
        int high = nums.length - 1;

        // Is there only one element or is array in sorted order?
        if (nums[0] <= nums[high]) {
            return nums[0];
        }

        while (low <= high) {
            int mid = (low + high) / 2;

            // Are we at rotation point?
            if (mid > 0 && nums[mid] < nums[mid - 1]) {
                return nums[mid];
            }
            if (mid < nums.length - 1 && nums[mid + 1] < nums[mid]) {
                return nums[mid + 1];
            }

            // Are we still before rotation point?
            if (nums[mid] > nums[0]) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return 0;
    }
}